pkg/tidydns/tidydns.go

```diff
@@ -93,6 +102,185 @@ func New(baseURL, username, password string) TidyDNSClient {
+		return nil, fmt.Errorf("subnet not found: %s", subnetCIDR)
+	}
+
+	if len(subnets) > 1 {
+		return nil, fmt.Errorf("too many subnets found: %s", subnetCIDR)
+	}
+
+	return &SubnetIDs{
+		SubnetID: subnets[0].ID,
+		ZoneID:   subnets[0].ZoneID,
+		VlanNo:   subnets[0].VlanNo,
+	}, nil
+}
+
+func (c *tidyDNSClient) GetFreeIP(ctx context.Context, subnetID int) (string, error) {
+	req, err := http.NewRequestWithContext(ctx, "GET", fmt.Sprintf("%s/=/dhcp_subnet_free_ip/%d", c.baseURL, subnetID), nil)
+	if err != nil {
+		return "", err
+	}
+
+	req.SetBasicAuth(c.username, c.password)
+
+	res, err := c.client.Do(req)
+	if err != nil {
+		return "", err
+	}
+	if res.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("error from tidyDNS server: %s", res.Status)
+	}
+
+	var freeIP dhcpFreeIP
+	err = json.NewDecoder(res.Body).Decode(&freeIP)
+	if err != nil {
+		return "", err
+	}
+
+	return freeIP.Data.IPAddress, nil
+}
+
+func (c *tidyDNSClient) CreateDHCPInterface(ctx context.Context, createInfo CreateInfo) (int, error) {
+	data := url.Values{
+		"subnet_id":   {strconv.Itoa(createInfo.SubnetID)},
+		"zone_id":     {strconv.Itoa(createInfo.ZoneID)},
+		"name":        {createInfo.InterfaceName},
+		"destination": {createInfo.InterfaceIP},
+		"location_id": {strconv.Itoa(createInfo.LocationID)},
+	}
+
+	var checkstring string = fmt.Sprintf("Key (destination)=(%s) already exists", createInfo.InterfaceIP)
+
+	req, err := http.NewRequestWithContext(ctx, "POST", fmt.Sprintf("%s/=/dhcp_interface//new", c.baseURL), strings.NewReader(data.Encode()))
+	if err != nil {
+		return 0, err
+	}
+	req.SetBasicAuth(c.username, c.password)
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	res, _ := c.client.Do(req)
+
+	if res.StatusCode != http.StatusOK {
+		bodyBytes, err := io.ReadAll(res.Body)
+		bodyString := string(bodyBytes)
+		if err != nil {
+			return 0, err
+		}
+		if strings.Contains(bodyString, checkstring) {
+			return 1, fmt.Errorf("try again")
+		}
+		return 0, fmt.Errorf("error from tidyDNS server: %s", res.Status)
+	}
+
+	var createResp interfaceCreate
+	err = json.NewDecoder(res.Body).Decode(&createResp)
+	if err != nil {
+		return 0, err
+	}
+
+	return createResp.ID, nil
+}
+
+func (c *tidyDNSClient) ReadDHCPInterface(ctx context.Context, interfaceID int) (*InterfaceInfo, error) {
+	req, err := http.NewRequestWithContext(ctx, "GET", fmt.Sprintf("%s/=/dhcp_interface/?id=%d", c.baseURL, interfaceID), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	req.SetBasicAuth(c.username, c.password)
+
+	res, err := c.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	if res.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("error from tidyDNS server: %s", res.Status)
+	}
+
+	var interfaceRead interfaceRead
+	err = json.NewDecoder(res.Body).Decode(&interfaceRead)
+	if err != nil {
+		return nil, err
+	}
+
+	return &InterfaceInfo{
+		InterfaceIP:   interfaceRead.Destination,
+		Interfacename: interfaceRead.Name,
+	}, nil
```

Comment:
Is there any reason for creating a whole new interfaceRead type when you are just copying the data to InterfaceInfo. Seems like you only need one or the other

Reply:
I agree it might be superfluous, but I would prefer not breaking backward compatibility by changing the interface.
Also I think it might be wise to separate the internals from the interface we expose to the users of the go api.